// rss3/dist/module/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const persona_1 = __importDefault(require("./persona"));
const file_1 = __importDefault(require("./file"));
const profile_1 = __importDefault(require("./profile"));
const items_1 = __importDefault(require("./items"));
const item_1 = __importDefault(require("./item"));
const links_1 = __importDefault(require("./links"));
const link_1 = __importDefault(require("./link"));
class RSS3 {
    constructor(options) {
        this.options = options;
        this.file = new file_1.default(this);
        this.persona = new persona_1.default(this);
        this.profile = new profile_1.default(this);
        this.items = new items_1.default(this);
        this.item = new item_1.default(this);
        this.links = new links_1.default(this);
        this.link = new link_1.default(this);
    }
}
exports.default = RSS3;
