// rss3/dist/module/link.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Link {
    constructor(main) {
        this.main = main;
    }
    async post(type, personaID) {
        const file = await this.main.file.get(this.main.persona.id);
        const lks = (file.links || []).find((links) => links.type === type);
        if (lks) {
            if (!lks.list) {
                lks.list = [];
            }
            const index = lks.list.indexOf(personaID);
            if (index === -1) {
                lks.list.push(personaID);
            }
            else {
                throw Error('Link already exist');
            }
            await this.main.file.set(file);
            return lks;
        }
        else {
            await this.main.links.post({
                type: type,
                list: [personaID],
            });
        }
    }
    async delete(type, personaID) {
        const file = await this.main.file.get(this.main.persona.id);
        const lks = (file.links || []).find((links) => links.type === type);
        if (lks) {
            if (!lks.list) {
                lks.list = [];
            }
            const index = lks.list.indexOf(personaID);
            if (index !== -1) {
                lks.list.splice(index, 1);
                if (lks.list.length === 0) {
                    delete lks.list;
                }
            }
            else {
                throw Error('Link does not exist');
            }
            await this.main.file.set(file);
            return lks;
        }
        else {
            throw Error('Link type does not exist');
        }
    }
}
exports.default = Link;
